rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadata.java

```diff
@@ -0,0 +1,309 @@
+ * composite.
+ */
+public final class CompositeMetadata {
+
+    /**
+     * An entry in a {@link CompositeMetadata}, which exposes the {@link #getMimeType() mime type} and {@link ByteBuf}
+     * {@link #getMetadata() content} of the metadata entry.
+     */
+    public interface Entry {
+
+        /**
+         * A <i>passthrough</i> entry is one for which the {@link #getMimeType()} could not be decoded.
+         * This is usually because it was compressed on the wire, but using an id that is still just "reserved for
+         * future use" in this implementation.
+         * <p>
+         * Still, another actor on the network might be able to interpret such an entry, which should thus be
+         * re-encoded as it was when forwarding the frame.
+         * <p>
+         * The {@link #getMetadata()} exposes the raw content buffer of the entry (as any other entry).
+         *
+         * @return true if this entry should be ignored but passed through as is during re-encoding
+         */
+        default boolean isPassthrough() {
+            return false;
+        }
+
+        /**
+         * @return the mime type for this entry
+         */
+        String getMimeType();
+
+        /**
+         * @return the metadata content of this entry
+         */
+        ByteBuf getMetadata();
+    }
+
+
+    /**
+     * Decode a {@link ByteBuf} into a {@link CompositeMetadata}. This is only possible on frame types used to initiate
+     * interactions, if the SETUP metadata mime type was {@link WellKnownMimeType#MESSAGE_RSOCKET_COMPOSITE_METADATA}.
+     * <p>
+     * Each entry {@link ByteBuf} is a {@link ByteBuf#readRetainedSlice(int) retained slice} of the original buffer.
+     *
+     * @param buffer the buffer to decode
+     * @return the decoded {@link CompositeMetadata}
+     */
+    public static CompositeMetadata decode(ByteBuf buffer) {
+        List<Entry> entries = new ArrayList<>();
+        while (buffer.isReadable()) {
+            entries.add(decodeIncrementally(buffer, true));
+        }
+        return new CompositeMetadata(entries);
+    }
+
+    /**
+     * Incrementally decode the next metadata entry from a {@link ByteBuf} into an {@link Entry}.
+     * This is only possible on frame types used to initiate
+     * interactions, if the SETUP metadata mime type was {@link WellKnownMimeType#MESSAGE_RSOCKET_COMPOSITE_METADATA}.
+     * <p>
+     * Each entry {@link ByteBuf} is a  {@link ByteBuf#readSlice(int) slice} of the original buffer that can also be
+     * {@link ByteBuf#readRetainedSlice(int) retained} if needed.
+     *
+     * @param buffer the buffer to decode
+     * @param retainMetadataSlices should each slide be retained when read from the original buffer?
+     * @return the decoded {@link Entry}
+     */
+    public static Entry decodeIncrementally(ByteBuf buffer, boolean retainMetadataSlices) {
+        Object[] entry = CompositeMetadataFlyweight.decodeNext(buffer, retainMetadataSlices);
```

Comment:
This should return ByteBufs, and the Entry type should be a flyweight. We shouldn't need to allocate any objects when we read this

Reply:
so `ByteBuf[2]`? in this case I'm thinking we can expose the mime buffer this way:
 - if the mime is an encoded id, a 1 byte long `ByteBuf` that represents the still encoded id (raw slice from the original buffer, need to do further bit manipulation)
 - if the mime is as custom string, a 2 to 129 bytes long `ByteBuf`, a slice of the underlying buffer that encompasses both the length byte and the mime string itself.

Since the mime string slide would have no more readable bytes than necessary to decode the string, the first byte could simply be discarded. Yet without it **there would be a corner case ambiguity for `1 byte` long buffer`**, which can either be:
 1. the case of a 1 byte encoded id
 2. the case of a 1 byte long String

So reading the mime buffer is a matter of checking its size is 1 byte and applying the mask OR size is 2+ bytes and skipping 1 byte then reading as `CharSequence`.

Does that sound good?

Reply:
note: the second `ByteBuf` would be a simple sliced of the original part that contains the metadata content, so its `readableBytes` reflect the size of the "content" (not including the encoded length)